js/block.js

```diff
@@ -143,55 +153,127 @@ StepProto.run = function(scope){
-        val.setAttribute('value', newVariableName);
-        val.textContent = newVariableName;
-    });
+    updateVariableNameInInstances(newVariableName, getVariablesToUpdate(context, parent.id));
 }
 
 function handleVariableBlur(evt){
     // Cleanup
     var input = evt.target;
-    if (!input.parentElement.hasAttribute('isvariable')){
-        return;
-    }
-    variableLocalsToUpdate = null;
+    var parent = dom.closest(input, '[script="control.setVariable"]');
+    ensureNameIsUniqueInContext(input);
     oldVariableName = '';
 }
 
-Event.on(document.body, 'editor:focus', '[isvariable] input', handleVariableFocus); // Mozilla
-Event.on(document.body, 'editor:focusin', '[isvariable] input', handleVariableFocus); // All other browsers
-Event.on(document.body, 'editor:input', '[isvariable] input', handleVariableInput);
-Event.on(document.body, 'editor:blur',  '[isvariable] input', handleVariableBlur); // Mozilla
-Event.on(document.body, 'editor:focusout',  '[isvariable] input', handleVariableBlur); // All other browsers
+function trailingNumber(str){
+    // return trailing number as a number or zero if no trailing number
+    return Number((str.match(/\d+$/) || [0])[0]);
+}
+
+function ensureNameIsUniqueInContext(input){
+    var parentContext = dom.closest(input, 'wb-contains');
+    var setVariable = dom.closest(input, '[script="control.setVariable"]');
+    var valueBlock = dom.closest(input, 'wb-value');
+    // Find other variable names in scope
+    var variablesToTestAgainst = dom.findAll(parentContext, '[script="control.setVariable"]')
+        .filter(function(setVarBlock){ return setVarBlock && setVarBlock !== setVariable; })
+        .map(function(setVarBlock){return dom.find(setVarBlock, 'input').value; })
+        .sort();
+    var newVariableName = input.value; // we may be changing this one
+    var oldVariableName = input.value;
+    // Compare against other variable names, update if there is a match
+    while(variablesToTestAgainst.indexOf(newVariableName) > -1){
+        var incrementalNumber = trailingNumber(newVariableName);
+        var baseName;
+        if (incrementalNumber){
+            baseName = newVariableName.slice(0, -(''+incrementalNumber).length); // trim off number
+        }else{
+            baseName = newVariableName + ' ';
+        }
+        newVariableName = baseName + (incrementalNumber + 1);
+    }
+    if (newVariableName !== oldVariableName){
+        console.log('old name: "%s", new name: "%s"', oldVariableName, newVariableName);
+        input.value = newVariableName;
+        valueBlock.setAttribute('value', newVariableName);
+        var variablesToUpdate = getVariablesToUpdate(parentContext, setVariable.id);
+        console.log('%s variables to update', variablesToUpdate.length);
+        if (variablesToUpdate.length){
+            console.log(variablesToUpdate[0]);
+        }
+        updateVariableNameInInstances(newVariableName, variablesToUpdate);
+    }
```

Comment:
This method is too long, refactor to be easier to follow and have the step manage its own internal state.